ipldutil/traverser.go

```diff
@@ -131,12 +132,20 @@ type traverser struct {
 	currentLink    ipld.Link
 	currentContext ipld.LinkContext
 	budget         *traversal.Budget
-	isDone         bool
+	isDone         bool // TODO(mvdan): we could dedup this with stopped
 	completionErr  error
-	awaitRequest   chan struct{}
-	stateChan      chan state
-	responses      chan nextResponse
-	stopped        chan struct{}
+
+	// TODO(mvdan): exactly what do these two chans do?
```

Comment:
Concur on the complexity cause I had to reread my code to rediscover the intent. But here it is:

From the standpoint of the consumer of the api, IsComplete and CurrentRequest give you information about the state of the traversal.

The state is cached 4 values:
- isDone = is everything wrapped up
- completionError - if it finished, did everything go ok?
- currentLink - what am I supposed to load next
- currentContext - waht is the link context for the next load

When I call `Advance` or `Error`, it sends a response to the loader which allows the traversal to continue. All of these values that are cached are no longer valid for the next call to IsComplete or CurrentRequest. Putting a value in awaitRequest essentially indicates on the next call to IsComplete or CurrentRequest, you need to wait and update the state. This is done by reading from the stateChan.

As I look at this, I agree that there are concurrency hickups here -- If the consumer interface were used by multiple threads for the same traversal there could be race conditions if two different threads called IsComplete/CurrentRequest around the same time when awaitRequest had a value. As it is, this hasn't come up yet cause there's only ever a single thread consuming, but we should fix it. I wonder if we should turn to my new favorite friend sync.Cond somehow (I've become way to enamored of it's Broadcast capabilites). But if there are good ways to simplify this period, I agree it's very confusing.

Reply:
Thanks for the extra context! It's all clear now.